// @syncfusion/ej2-grids/src/grid/services/intersection-observer.js
import { EventHandler, Browser, isNullOrUndefined } from '@syncfusion/ej2-base';
import { debounce } from '@syncfusion/ej2-base';
/**
 * InterSectionObserver - class watch whether it enters the viewport.
 * @hidden
 */
var InterSectionObserver = /** @class */ (function () {
    function InterSectionObserver(parent, element, options) {
        var _this = this;
        this.fromWheel = false;
        this.touchMove = false;
        this.options = {};
        this.sentinelInfo = {
            'up': {
                check: function (rect, info) {
                    var top = rect.top - _this.containerRect.top;
                    info.entered = top >= 0;
                    return top + (_this.options.pageHeight / 2) >= 0;
                },
                axis: 'Y'
            },
            'down': {
                check: function (rect, info) {
                    var cHeight = _this.options.container.clientHeight;
                    var top = rect.bottom;
                    info.entered = rect.bottom <= _this.containerRect.bottom;
                    return top - (_this.options.pageHeight / 2) <= _this.options.pageHeight / 2;
                }, axis: 'Y'
            },
            'right': {
                check: function (rect, info) {
                    var right = rect.right;
                    info.entered = right < _this.containerRect.right;
                    return right - _this.containerRect.width <= _this.containerRect.right;
                }, axis: 'X'
            },
            'left': {
                check: function (rect, info) {
                    var left = rect.left;
                    info.entered = left > 0;
                    return left + _this.containerRect.width >= _this.containerRect.left;
                }, axis: 'X'
            }
        };
        this.parent = parent;
        this.element = element;
        this.options = options;
        this.addEventListener();
    }
    InterSectionObserver.prototype.virtualKeyHandler = function (args) {
        if (args && (args.action === 'upArrow' || args.action === 'downArrow') && isNullOrUndefined(this.activeKey)) {
            this.activeKey = args.action;
        }
    };
    /**
     * @hidden
     */
    InterSectionObserver.prototype.addEventListener = function () {
        this.parent.on('virtaul-key-handler', this.virtualKeyHandler, this);
        this.parent.on('key-pressed', this.virtualKeyHandler, this);
    };
    /**
     * @hidden
     */
    InterSectionObserver.prototype.removeEventListener = function () {
        if (this.parent.isDestroyed) {
            return;
        }
        this.parent.off('virtaul-key-handler', this.virtualKeyHandler);
        this.parent.off('key-pressed', this.virtualKeyHandler);
    };
    /**
     * @hidden
     */
    InterSectionObserver.prototype.destroy = function () {
        this.removeEventListener();
    };
    InterSectionObserver.prototype.observe = function (callback, onEnterCallback) {
        var _this = this;
        this.containerRect = this.options.container.getBoundingClientRect();
        EventHandler.add(this.options.container, 'wheel', function () { return _this.fromWheel = true; }, this);
        EventHandler.add(this.options.container, 'scroll', this.virtualScrollHandler(callback, onEnterCallback), this);
    };
    InterSectionObserver.prototype.check = function (direction) {
        var info = this.sentinelInfo[direction];
        return info.check(this.element.getBoundingClientRect(), info);
    };
    InterSectionObserver.prototype.ensureKeyFocus = function () {
        if (this.activeKey !== 'upArrow' && this.activeKey !== 'downArrow') {
            return false;
        }
        var row = document.activeElement.parentElement;
        var rowIndex = parseInt(row.getAttribute('aria-rowindex'), 10);
        var blockSize = this.parent.contentModule.getBlockSize() - 1;
        rowIndex = this.activeKey === 'upArrow' ? rowIndex - blockSize : rowIndex + blockSize;
        return !isNullOrUndefined(this.parent.getRowByIndex(rowIndex));
    };
    InterSectionObserver.prototype.virtualScrollHandler = function (callback, onEnterCallback) {
        var _this = this;
        var delay = Browser.info.name === 'chrome' ? 200 : 100;
        var prevTop = 0;
        var prevLeft = 0;
        var debounced100 = debounce(callback, delay);
        var debounced50 = debounce(callback, 50);
        return function (e) {
            var isKeyDown = _this.ensureKeyFocus();
            if (!isKeyDown) {
                var top_1 = e.target.scrollTop;
                var left = e.target.scrollLeft;
                var direction = prevTop < top_1 ? 'down' : 'up';
                direction = prevLeft === left ? direction : prevLeft < left ? 'right' : 'left';
                prevTop = top_1;
                prevLeft = left;
                var current = _this.sentinelInfo[direction];
                if (_this.options.axes.indexOf(current.axis) === -1) {
                    return;
                }
                var check = _this.check(direction);
                if (current.entered) {
                    onEnterCallback(_this.element, current, direction, { top: top_1, left: left }, _this.fromWheel, check);
                }
                if (check) {
                    var fn = debounced100;
                    //this.fromWheel ? this.options.debounceEvent ? debounced100 : callback : debounced100;
                    if (current.axis === 'X') {
                        fn = debounced50;
                    }
                    fn({ direction: direction, sentinel: current, offset: { top: top_1, left: left },
                        focusElement: document.activeElement });
                }
                _this.fromWheel = false;
            }
            else {
                _this.activeKey = undefined;
            }
        };
    };
    InterSectionObserver.prototype.setPageHeight = function (value) {
        this.options.pageHeight = value;
    };
    return InterSectionObserver;
}());
export { InterSectionObserver };
