// @syncfusion/ej2-kanban/src/kanban/base/css-constant.js
/**
 * Kanban CSS Constants
 */
/** @hidden */
export var ROOT_CLASS = 'e-kanban';
/** @hidden */
export var RTL_CLASS = 'e-rtl';
/** @hidden */
export var DEVICE_CLASS = 'e-device';
/** @hidden */
export var ICON_CLASS = 'e-icons';
/** @hidden */
export var TEMPLATE_CLASS = 'e-template';
/** @hidden */
export var SWIMLANE_CLASS = 'e-swimlane';
/** @hidden */
export var TABLE_CLASS = 'e-kanban-table';
/** @hidden */
export var HEADER_CLASS = 'e-kanban-header';
/** @hidden */
export var HEADER_TABLE_CLASS = 'e-header-table';
/** @hidden */
export var HEADER_CELLS_CLASS = 'e-header-cells';
/** @hidden */
export var HEADER_WRAP_CLASS = 'e-header-wrap';
/** @hidden */
export var HEADER_TEXT_CLASS = 'e-header-text';
/** @hidden */
export var HEADER_ICON_CLASS = 'e-header-icon';
/** @hidden */
export var CONTENT_CELLS_CLASS = 'e-content-cells';
/** @hidden */
export var CONTENT_CLASS = 'e-kanban-content';
/** @hidden */
export var CONTENT_TABLE_CLASS = 'e-content-table';
/** @hidden */
export var HEADER_ROW_TOGGLE_CLASS = 'e-toggle-header';
/** @hidden */
export var HEADER_ROW_CLASS = 'e-header-row';
/** @hidden */
export var CONTENT_ROW_CLASS = 'e-content-row';
/** @hidden */
export var SWIMLANE_ROW_CLASS = 'e-swimlane-row';
/** @hidden */
export var SWIMLANE_ROW_EXPAND = 'e-swimlane-row-expand';
/** @hidden */
export var SWIMLANE_ROW_COLLAPSE = 'e-swimlane-row-collapse';
/** @hidden */
export var SWIMLANE_ROW_TEXT = 'e-swimlane-text';
/** @hidden */
export var SWIMLANE_ITEM_COUNT = 'e-swimlane-item-count';
/** @hidden */
export var CARD_WRAPPER_CLASS = 'e-card-wrapper';
/** @hidden */
export var CARD_CLASS = 'e-card';
/** @hidden */
export var CARD_HEADER_CLASS = 'e-card-header';
/** @hidden */
export var CARD_CONTENT_CLASS = 'e-card-content';
/** @hidden */
export var CARD_HEADER_TEXT_CLASS = 'e-card-header-caption';
/** @hidden */
export var CARD_CONTENT_TEXT_CLASS = 'e-card-content-caption';
/** @hidden */
export var COLUMN_EXPAND = 'e-column-expand';
/** @hidden */
export var COLUMN_COLLAPSE = 'e-column-collapse';
/** @hidden */
export var COLLAPSE_HEADER_TEXT = 'e-collapse-header-text';
/** @hidden */
export var COLLAPSED_CLASS = 'e-collapsed';
/** @hidden */
export var DIALOG_CLASS = 'e-kanban-dialog';
/** @hidden */
export var DIALOG_CONTENT_CONTAINER = 'e-kanban-dialog-content';
/** @hidden */
export var FORM_CLASS = 'e-kanban-form';
/** @hidden */
export var FORM_WRAPPER_CLASS = 'e-kanban-form-wrapper';
/** @hidden */
export var ERROR_VALIDATION_CLASS = 'e-kanban-error';
/** @hidden */
export var FIELD_CLASS = 'e-field';
/** @hidden */
export var DRAGGED_CLONE = 'e-target-dragged-clone';
/** @hidden */
export var DRAGGED_CARD = 'e-kanban-dragged-card';
/** @hidden */
export var DROPPED_CLONE = 'e-target-dropped-clone';
/** @hidden */
export var DROPPING = 'e-dropping';
/** @hidden */
export var TOGGLE_VISIBLE = 'e-toggle-visible';
/** @hidden */
export var MULTI_CLONE_CARD = 'e-multi-clone-card';
/** @hidden */
export var MULTI_CLONE_CONTENT_CELL = 'e-multi-content-cell';
/** @hidden */
export var MULTI_ACTIVE = 'e-multi-active';
/** @hidden */
export var TARGET_MULTI_CLONE = 'e-target-multi-clone';
/** @hidden */
export var MULTI_COLUMN_KEY = 'e-column-key';
/** @hidden */
export var CARD_SELECTION = 'e-selection';
/** @hidden */
export var TOOLTIP_CLASS = 'e-kanban-tooltip';
/** @hidden */
export var TOOLTIP_TEXT = 'e-tooltip-text';
/** @hidden */
export var SCROLL_CONTAINER_CLASS = 'e-cell-scroll-container';
/** @hidden */
export var SWIMLANE_HEADER = 'e-swimlane-header';
/** @hidden */
export var SWIMLANE_HEADER_TOOLBAR = 'e-swimlane-header-toolbar';
/** @hidden */
export var TOOLBAR_MENU = 'e-toolbar-menu';
/** @hidden */
export var TOOLBAR_MENU_ICON = 'e-icon-menu';
/** @hidden */
export var TOOLBAR_LEVEL_TITLE = 'e-toolbar-level-title';
/** @hidden */
export var TOOLBAR_SWIMLANE_NAME = 'e-toolbar-swimlane-name';
/** @hidden */
export var SWIMLANE_OVERLAY = 'e-swimlane-overlay';
/** @hidden */
export var SWIMLANE_CONTENT = 'e-swimlane-content';
/** @hidden */
export var SWIMLANE_RESOURCE = 'e-swimlane-resource';
/** @hidden */
export var SWIMLANE_TREE = 'e-swimlane-tree';
/** @hidden */
export var LIMITS_CLASS = 'e-limits';
/** @hidden */
export var MAX_COUNT_CLASS = 'e-max-count';
/** @hidden */
export var MIN_COUNT_CLASS = 'e-min-count';
/** @hidden */
export var TOTAL_CARD = 'e-total-card';
/** @hidden */
export var HIDE_LIMITS = 'e-hide-limits';
/** @hidden */
export var MAX_COLOR = 'e-max-color';
/** @hidden */
export var MIN_COLOR = 'e-min-color';
